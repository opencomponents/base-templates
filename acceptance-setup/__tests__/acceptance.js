/**
 * @testEnvironment jsdom
 */

jasmine.DEFAULT_TIMEOUT_INTERVAL = 100000;

const server = require('../server');
const { cli, Registry } = require('oc');
const path = require('path');
const r = require('request-promise-native');
const jsdom = require('jsdom');
const JSDOM = jsdom.JSDOM;
const fs = require('fs-extra');

const registryPort = 3000;
const registryUrl = `http://localhost:${registryPort}/`;
const serverPort = 4000;
const serverUrl = `http://localhost:${serverPort}/`;
const components = [
  {
    name: 'base-component-handlebars',
    template: require('../../packages/oc-template-handlebars'),
    path: path.join(
      __dirname,
      '../../acceptance-components/base-component-handlebars'
    )
  },
  {
    name: 'base-component-jade',
    template: require('../../packages/oc-template-jade'),
    path: path.join(
      __dirname,
      '../../acceptance-components/base-component-jade'
    )
  },
  {
    name: 'base-component-es6',
    template: require('../../packages/oc-template-es6'),
    path: path.join(__dirname, '../../acceptance-components/base-component-es6')
  }
];

let registry;
let testServer;

const semverRegex = /\bv?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-[\da-z-]+(?:\.[\da-z-]+)*)?(?:\+[\da-z-]+(?:\.[\da-z-]+)*)?\b/gi;

beforeAll(done => {
  components.forEach(component => {
    fs.removeSync(path.join(component.path, '_package'));
  });

  const promisify = fn => options =>
    new Promise((resolve, reject) =>
      fn(options, err => (err ? reject(err) : resolve()))
    );
  const packageAsync = promisify(cli.package);

  const packaged = components.map(component =>
    packageAsync({
      componentPath: component.path
    })
  );

  Promise.all(packaged)
    .then(() => {
      registry = new Registry({
        local: true,
        discovery: true,
        verbosity: 0,
        path: path.join(__dirname, '../../acceptance-components'),
        port: registryPort,
        baseUrl: registryUrl,
        env: {},
        templates: components
          .map(component => component.template)
          .filter(template => {
            // we need only the non-default templates here
            const type = template.getInfo().type;
            return (
              type !== 'oc-template-handlebars' && type !== 'oc-template-jade'
            );
          })
      });

      registry.start(err => {
        if (err) {
          return done(err);
        }
        testServer = server(serverPort, err => {
          if (err) {
            return done(err);
          }
          done();
        });
      });
    })
    .catch(err => {
      return done(err);
    });
});

afterAll(done => {
  testServer.close(() => {
    registry.close(() => {
      components.forEach(component => {
        fs.removeSync(path.join(component.path, '_package'));
      });
      done();
    });
  });
});

test('Registry should correctly serve rendered components', done => {
  const rendered = components.map(component =>
    r(registryUrl + `${component.name}/?name=SuperMario`)
      .then(body => {
        const bodyVersionless = body
          .replace(semverRegex, '6.6.6')
          .replace(/data-hash=\\\".*?\\\"/, '')
          .replace(
            /\[\\\"oc\\\",.*?\\\"reactComponents\\\",.*?\\\".*?\\\"\]/,
            '["oc", "reactComponents", "dummyContent"]'
          );
        return Promise.resolve(bodyVersionless);
      })
      .catch(err => Promise.reject(err))
  );
  Promise.all(rendered)
    .then(responses => {
      responses.forEach(response => expect(response).toMatchSnapshot());
      done();
    })
    .catch(err => done(err));
});

test('Registry should correctly serve unrendered components', done => {
  const unrendered = components.map(component =>
    r({
      uri: registryUrl + `${component.name}/?name=SuperMario`,
      headers: {
        Accept: 'application/vnd.oc.unrendered+json'
      }
    })
      .then(body => {
        const bodyVersionless = body
          .replace(semverRegex, '6.6.6')
          .replace(/\"key\":\".*?\"/g, '');
        return Promise.resolve(bodyVersionless);
      })
      .catch(err => Promise.reject(err))
  );

  Promise.all(unrendered)
    .then(responses => {
      responses.forEach(response => expect(response).toMatchSnapshot());
      done();
    })
    .catch(err => done(err));
});

test('server-side-side rendering', done => {
  JSDOM.fromURL(`${serverUrl}?name=SuperMario`, {})
    .then(dom => {
      const domVersionless = dom.serialize().replace(semverRegex, '6.6.6');
      expect(domVersionless).toMatchSnapshot();
      done();
    })
    .catch(err => done(err));
});

test('client-side-side rendering', done => {
  const rendered = components.map(component =>
    JSDOM.fromURL(`${registryUrl}${component.name}/~preview?name=SuperMario`, {
      resources: 'usable',
      runScripts: 'dangerously'
    })
      .then(
        dom =>
          new Promise(resolve => {
            setTimeout(() => {
              const node = dom.window.document.getElementById(component.name);
              return resolve(node);
            }, 5000);
          })
      )
      .catch(err => Promise.reject(err))
  );

  Promise.all(rendered)
    .then(responses => {
      responses.forEach(response => expect(response).toMatchSnapshot());
      done();
    })
    .catch(err => done(err));
});
