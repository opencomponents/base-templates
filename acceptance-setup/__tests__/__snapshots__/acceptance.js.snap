// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry should correctly serve rendered components 1`] = `"{\\"type\\":\\"oc-component-local\\",\\"version\\":\\"6.6.6\\",\\"requestVersion\\":\\"\\",\\"name\\":\\"base-component-handlebars\\",\\"renderMode\\":\\"rendered\\",\\"href\\":\\"http://localhost:3000/base-component-handlebars?name=SuperMario\\",\\"html\\":\\"<oc-component href=\\\\\\"http://localhost:3000/base-component-handlebars?name=SuperMario\\\\\\"  data-name=\\\\\\"base-component-handlebars\\\\\\" data-rendered=\\\\\\"true\\\\\\" data-version=\\\\\\"6.6.6\\\\\\"><div id=\\\\\\"base-component-handlebars\\\\\\">\\\\n  <img src=\\\\\\"http://localhost:3000/base-component-handlebars/6.6.6/static/img/logo.png\\\\\\" />\\\\n  <div>Hello SuperMario</div>\\\\n</div>\\\\n\\\\n<script>window.oc=window.oc||{};oc.renderedComponents=oc.renderedComponents||{};oc.renderedComponents[\\\\\\"base-component-handlebars\\\\\\"]=\\\\\\"6.6.6\\\\\\";</script></oc-component>\\"}"`;

exports[`Registry should correctly serve rendered components 2`] = `"{\\"type\\":\\"oc-component-local\\",\\"version\\":\\"6.6.6\\",\\"requestVersion\\":\\"\\",\\"name\\":\\"base-component-jade\\",\\"renderMode\\":\\"rendered\\",\\"href\\":\\"http://localhost:3000/base-component-jade?name=SuperMario\\",\\"html\\":\\"<oc-component href=\\\\\\"http://localhost:3000/base-component-jade?name=SuperMario\\\\\\"  data-name=\\\\\\"base-component-jade\\\\\\" data-rendered=\\\\\\"true\\\\\\" data-version=\\\\\\"6.6.6\\\\\\"><div id=\\\\\\"base-component-jade\\\\\\"><img src=\\\\\\"http://localhost:3000/base-component-jade/6.6.6/static/img/logo.png\\\\\\"/><div>Hello SuperMario</div></div><script>window.oc=window.oc||{};oc.renderedComponents=oc.renderedComponents||{};oc.renderedComponents[\\\\\\"base-component-jade\\\\\\"]=\\\\\\"6.6.6\\\\\\";</script></oc-component>\\"}"`;

exports[`Registry should correctly serve rendered components 3`] = `"{\\"type\\":\\"oc-component-local\\",\\"version\\":\\"6.6.6\\",\\"requestVersion\\":\\"\\",\\"name\\":\\"base-component-es6\\",\\"renderMode\\":\\"rendered\\",\\"href\\":\\"http://localhost:3000/base-component-es6?name=SuperMario\\",\\"html\\":\\"<oc-component href=\\\\\\"http://localhost:3000/base-component-es6?name=SuperMario\\\\\\"  data-name=\\\\\\"base-component-es6\\\\\\" data-rendered=\\\\\\"true\\\\\\" data-version=\\\\\\"6.6.6\\\\\\"><div id=\\\\\\"base-component-es6\\\\\\" class=awesome__2cThMnJY>\\\\n Hello SUPERMARIO!\\\\n <img src=\\\\\\"http://localhost:3000/base-component-es6/6.6.6/static/img/logo.png\\\\\\" />\\\\n </div><style>.awesome__2cThMnJY{font-family:\\\\\\"Hiragino Sans\\\\\\",\\\\\\"ヒラギノ角ゴシック\\\\\\",\\\\\\"Hiragino Kaku Gothic ProN\\\\\\",\\\\\\"ヒラギノ角ゴ ProN W3\\\\\\",Roboto,\\\\\\"Droid Sans\\\\\\",\\\\\\"游ゴシック\\\\\\",YuGothic,\\\\\\"メイリオ\\\\\\",Meiryo,sans-serif;background:#db7093}</style><script>window.oc = window.oc || {};oc.cmd = oc.cmd || [];oc.cmd.push(function(oc){oc.events.fire('oc:cssDidMount', '.awesome__2cThMnJY{font-family:\\\\\\"Hiragino Sans\\\\\\",\\\\\\"ヒラギノ角ゴシック\\\\\\",\\\\\\"Hiragino Kaku Gothic ProN\\\\\\",\\\\\\"ヒラギノ角ゴ ProN W3\\\\\\",Roboto,\\\\\\"Droid Sans\\\\\\",\\\\\\"游ゴシック\\\\\\",YuGothic,\\\\\\"メイリオ\\\\\\",Meiryo,sans-serif;background:#db7093}');});</script><script>window.oc=window.oc||{};oc.renderedComponents=oc.renderedComponents||{};oc.renderedComponents[\\\\\\"base-component-es6\\\\\\"]=\\\\\\"6.6.6\\\\\\";</script></oc-component>\\"}"`;

exports[`Registry should correctly serve unrendered components 1`] = `"{\\"type\\":\\"oc-component-local\\",\\"version\\":\\"6.6.6\\",\\"requestVersion\\":\\"\\",\\"name\\":\\"base-component-handlebars\\",\\"renderMode\\":\\"unrendered\\",\\"href\\":\\"http://localhost:3000/base-component-handlebars?name=SuperMario\\",\\"data\\":{\\"name\\":\\"SuperMario\\",\\"staticPath\\":\\"http://localhost:3000/base-component-handlebars/6.6.6/static/\\"},\\"template\\":{\\"src\\":\\"http://localhost:3000/base-component-handlebars/6.6.6/static/template.js\\",\\"type\\":\\"oc-template-handlebars\\",}}"`;

exports[`Registry should correctly serve unrendered components 2`] = `"{\\"type\\":\\"oc-component-local\\",\\"version\\":\\"6.6.6\\",\\"requestVersion\\":\\"\\",\\"name\\":\\"base-component-jade\\",\\"renderMode\\":\\"unrendered\\",\\"href\\":\\"http://localhost:3000/base-component-jade?name=SuperMario\\",\\"data\\":{\\"name\\":\\"SuperMario\\",\\"staticPath\\":\\"http://localhost:3000/base-component-jade/6.6.6/static/\\"},\\"template\\":{\\"src\\":\\"http://localhost:3000/base-component-jade/6.6.6/static/template.js\\",\\"type\\":\\"oc-template-jade\\",}}"`;

exports[`Registry should correctly serve unrendered components 3`] = `"{\\"type\\":\\"oc-component-local\\",\\"version\\":\\"6.6.6\\",\\"requestVersion\\":\\"\\",\\"name\\":\\"base-component-es6\\",\\"renderMode\\":\\"unrendered\\",\\"href\\":\\"http://localhost:3000/base-component-es6?name=SuperMario\\",\\"data\\":{\\"name\\":\\"SuperMario\\",\\"staticPath\\":\\"http://localhost:3000/base-component-es6/6.6.6/static/\\"},\\"template\\":{\\"src\\":\\"http://localhost:3000/base-component-es6/6.6.6/static/template.js\\",\\"type\\":\\"oc-template-es6\\",}}"`;

exports[`client-side-side rendering 1`] = `
<div
  id="base-component-handlebars"
>
  
  
  <img
    src="http://localhost:3000/base-component-handlebars/1.0.3/static/img/logo.png"
  />
  
  
  <div>
    Hello SuperMario
  </div>
  

</div>
`;

exports[`client-side-side rendering 2`] = `
<div
  id="base-component-jade"
>
  <img
    src="http://localhost:3000/base-component-jade/1.0.3/static/img/logo.png"
  />
  <div>
    Hello SuperMario
  </div>
</div>
`;

exports[`client-side-side rendering 3`] = `
<div
  class="awesome__2cThMnJY"
  id="base-component-es6"
>
  
 Hello SUPERMARIO!
 
  <img
    src="http://localhost:3000/base-component-es6/1.0.1/static/img/logo.png"
  />
  
 
</div>
`;

exports[`server-side-side rendering 1`] = `
"<!DOCTYPE html><html><head>
            <title>A page</title>
          </head>
          <body>
            <div id=\\"base-component-es6\\" class=\\"awesome__2cThMnJY\\">
 Hello SUPERMARIO!
 <img src=\\"http://localhost:3000/base-component-es6/6.6.6/static/img/logo.png\\">
 </div><style>.awesome__2cThMnJY{font-family:\\"Hiragino Sans\\",\\"ヒラギノ角ゴシック\\",\\"Hiragino Kaku Gothic ProN\\",\\"ヒラギノ角ゴ ProN W3\\",Roboto,\\"Droid Sans\\",\\"游ゴシック\\",YuGothic,\\"メイリオ\\",Meiryo,sans-serif;background:#db7093}</style><script>window.oc = window.oc || {};oc.cmd = oc.cmd || [];oc.cmd.push(function(oc){oc.events.fire('oc:cssDidMount', '.awesome__2cThMnJY{font-family:\\"Hiragino Sans\\",\\"ヒラギノ角ゴシック\\",\\"Hiragino Kaku Gothic ProN\\",\\"ヒラギノ角ゴ ProN W3\\",Roboto,\\"Droid Sans\\",\\"游ゴシック\\",YuGothic,\\"メイリオ\\",Meiryo,sans-serif;background:#db7093}');});</script><script>window.oc=window.oc||{};oc.renderedComponents=oc.renderedComponents||{};oc.renderedComponents[\\"base-component-es6\\"]=\\"6.6.6\\";</script>
            <div id=\\"base-component-handlebars\\">
  <img src=\\"http://localhost:3000/base-component-handlebars/6.6.6/static/img/logo.png\\">
  <div>Hello SuperMario</div>
</div>

<script>window.oc=window.oc||{};oc.renderedComponents=oc.renderedComponents||{};oc.renderedComponents[\\"base-component-handlebars\\"]=\\"6.6.6\\";</script>
            <div id=\\"base-component-jade\\"><img src=\\"http://localhost:3000/base-component-jade/6.6.6/static/img/logo.png\\"><div>Hello SuperMario</div></div><script>window.oc=window.oc||{};oc.renderedComponents=oc.renderedComponents||{};oc.renderedComponents[\\"base-component-jade\\"]=\\"6.6.6\\";</script>
            <script>window.oc=window.oc||{};oc.conf=oc.conf||{};oc.conf.templates=(oc.conf.templates||[]).concat([{\\"type\\":\\"oc-template-es6\\",\\"version\\":\\"6.6.6\\",\\"externals\\":[]},{\\"type\\":\\"oc-template-jade\\",\\"version\\":\\"6.6.6\\",\\"externals\\":[{\\"global\\":\\"jade\\",\\"url\\":\\"https://unpkg.com/jade-legacy@6.6.6/runtime.js\\",\\"name\\":\\"jade\\"}]},{\\"type\\":\\"oc-template-handlebars\\",\\"version\\":\\"6.6.6\\",\\"externals\\":[{\\"global\\":\\"Handlebars\\",\\"url\\":\\"https://unpkg.com/handlebars@6.6.6/dist/handlebars.runtime.min.js\\",\\"name\\":\\"handlebars\\"}]}]);</script><script src=\\"http://localhost:3000/oc-client/6.6.6/static/src/oc-client.min.js\\" type=\\"text/javascript\\"></script><script>window.oc=window.oc||{};oc.renderedComponents=oc.renderedComponents||{};oc.renderedComponents[\\"oc-client\\"]=\\"6.6.6\\";</script>
          
        
      </body></html>"
`;
